### PR TITLE
cockroach-go/testserver: testserver using real binaries

### PR DESCRIPTION
**Pull request by mberhault**
After discussion, moving to this repo.
I've also cleaned up node.go a log and collapsed everything into
a single struct (TestServer).

Among other things:
* dump logs on unexpected failures (non-zero exits before Stop())
* cleanup temp dir

Still todo:
* download binary automatically
* use unix socket when (and if) supported by the server

### CODE REVIEW

**Comment by mberhault** [comment]
last commit automatically downloads the latest binary for the current OS/ARCH. this is mostly error handling, downloading is hard!

**Comment by mberhault** [comment]
@bdarnell: this is the new naming + aws download.
I've tested it locally in https://github.com/cockroachdb/examples-go/pull/50, and it's quite happy.

File: testserver/binaries.go [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+	"os"
+	"path/filepath"
+	"runtime"
+	"strings"
+)
+
+const (
+	awsBaseURL      = "https://s3.amazonaws.com/cockroach/cockroach"
+	latestSuffix    = "LATEST"
+	localBinaryPath = "/var/tmp"
```
Comment (spencerkimball) [inline testserver/binaries.go]: Do we want to create a subdir for this stuff in `/var/tmp`?

File: testserver/testserver.go [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+	return ts.stderrBuf.String()
+}
+
+// PGURL parses the node's stdout log for the postgres URL.
+// It returns an actual URL so that callers may modify individual settings.
+// eg:
+//   u := server.PGURL()
+//   u.Path = "mydatabase"
+//   db, err := sql.Open("postgres", u.String())
+// NOTE: it may not be available for a few milliseconds after start,
```
Comment (spencerkimball) [inline testserver/testserver.go]: Let's remove this comment and not return until after we've gotten the port from `stdout`. Callers to this method should just block.

File: testserver/testserver_test.go [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+func TestRunServer(t *testing.T) {
+	ts, err := NewTestServer()
+	if err != nil {
+		t.Fatal(err)
+	}
+	ts.Start()
+	defer ts.Stop()
+
+	time.Sleep(time.Second)
```
Comment (spencerkimball) [inline testserver/testserver_test.go]: Remove this `Sleep` invocation.

**Comment by spencerkimball** [comment]
LGTM

File: testserver/testserver.go (hunk 2 above)
Comment (mberhault) [inline testserver/testserver.go]: added WaitForPGURL to loop over stdout and adjusted comment.

File: testserver/testserver_test.go (hunk 3 above)
Comment (mberhault) [inline testserver/testserver_test.go]: replaced with call to WaitForPGURL

File: testserver/binaries.go (hunk 1 above)
Comment (mberhault) [inline testserver/binaries.go]: I'm not sure we need to, though I don't know what the conventions are for `/var/tmp`. Can't really find much.

File: testserver/testserver.go (hunk 2 above)
Comment (spencerkimball) [inline testserver/testserver.go]: Why add the additional method? Just simplify and have this one which should use a mutex to block and then cache the URL result once it's computed.

Comment (mberhault) [inline testserver/testserver.go]: changed to a goroutine looking for the url instead and stores it permanently.

**Comment by spencerkimball** [comment]
LGTM

**Pull request merged**